google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkPreconditionsSupport.java

```diff
@@ -0,0 +1,199 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.appengine.sdk;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import com.google.cloud.tools.intellij.appengine.sdk.CloudSdkService.SdkStatus;
+import com.google.cloud.tools.intellij.flags.PropertiesFileFlagReader;
+import com.google.cloud.tools.intellij.util.GctBundle;
+import com.google.common.annotations.VisibleForTesting;
+import com.intellij.notification.Notification;
+import com.intellij.notification.NotificationType;
+import com.intellij.notification.Notifications;
+import com.intellij.openapi.actionSystem.AnAction;
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.options.ShowSettingsUtil;
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.openapi.progress.Task;
+import com.intellij.openapi.project.Project;
+import com.intellij.remoteServer.runtime.deployment.ServerRuntimeInstance.DeploymentOperationCallback;
+import com.intellij.remoteServer.runtime.log.LoggingHandler;
+import java.util.concurrent.CountDownLatch;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Utility class providing support for blocking / tracking {@link CloudSdkService#install()} process
+ * and other SDK preconditions so that dependent deployment processes can be postponed until SDK is
+ * completely ready.
+ */
+public class CloudSdkPreconditionsSupport {
+  private static CloudSdkPreconditionsSupport instance;
+
+  public static CloudSdkPreconditionsSupport getInstance() {
+    if (instance == null) {
+      instance = new CloudSdkPreconditionsSupport();
+    }
+
+    return instance;
+  }
+
+  /**
+   * Waits for Cloud SDK to be ready for all operations and then runs the given runnable. If process
+   * results in error or user cancel, shows notification and does not run.
+   *
+   * @param project Project to which runnable belongs.
+   * @param runnable Runnable to run after Cloud SDK is ready.
+   * @param deploymentLog Log to print SDK statuses.
+   * @param callback Callback to report errors.
+   */
+  public void runAfterCloudSdkPreconditionsMet(
+      @Nullable Project project,
+      Runnable runnable,
+      LoggingHandler deploymentLog,
+      DeploymentOperationCallback callback) {
```

Comment:
seems strange that this is tied to a `DeploymentOperationCallback` when its meant to be a generic installation manager. what about cases like blocking for local run?

Reply:
also, and I may be off, but could the logic in this class (for blocking operations until install complete) be folded into the `CloudSdkServiceManager`? Seems like this would be a natural place for it.

Reply:
I agree, I thought about putting most of the supportive functionality into a new structure of interface, something like `CloudSdkPreconditionCallback` since we don't know anything about deployment, runnable or logger, but decided it'd be easier to see what exactly goes into that method so the changes are easier to digest. If you think that makes sense, I'll refactor that. Open to suggestions on callback name too.

Reply:
Good idea about `CloudSdkServiceManager`, I'll do that.

Reply:
regarding the callback - doesn't it have to be something generic? What about blocking for local-run - don't we need that implemented before we mark this complete (and permanently enable the feature flag)?

Reply:
I think local run goes some other path, let me check it.

Reply:
Ah yes, I missed `CloudSdkStartupPolicy` where the checks are done too before running on local app dev server. Let me include that into the new flow as well.